File: 34. Find First and Last Position of Element in Sorted Array.java
class Solution {
    
    private int extremeInsertionIndex(int[] nums, int target, boolean left) {
        int lo = 0;
        int hi = nums.length; // cannot use nums.length - 1
        
        while (lo < hi) {
            int mid = (lo + hi)/2;
            if (nums[mid] > target || (left && nums[mid] == target)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo; 
    }
    
    public int[] searchRange(int[] nums, int target) {
        int[] targetRange = {-1, -1};
        int leftIdx = extremeInsertionIndex(nums,target,true);
        if (leftIdx == nums.length || nums[leftIdx] != target) {
            return targetRange;
        }
        
        int rightIdx = extremeInsertionIndex(nums,target,false) - 1; // because rightIdx -1 here 
        targetRange[0] = leftIdx;
        targetRange[1] = rightIdx;
        return targetRange;
        
    }
}

// Time: O(logn)
// Space: O(1)

File: amazon/python/34. Find First and Last Position of Element in Sorted Array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if len(nums) == 0:
            return [-1,-1]
        
        l = 0
        r = len(nums)
        
        res = [-1, -1]
        
        while l < r:
            mid = l + (r-l)//2
            if target > nums[mid]:
                l = mid + 1
            else:
                r = mid
        
        if r < len(nums) and r >= 0 and nums[r] == target: 
            res[0] = r
        else:
            return [-1,-1]
        
        l = 0
        r = len(nums)
        
        while l < r:
            mid = l + (r-l)//2
            if target >= nums[mid]:
                l = mid + 1
            else:
                r = mid
                
        res[1] = r - 1
        
        return res
    
    ## Time: O(logn)
    ## Space: O(1)
